### PR TITLE
[docs] A guide covering communication between native and RN on IOS.

### PR DESCRIPTION
**Pull request by sienkowski**
WARNING: I was not able to run the website locally because of a bug, so please test it before merging.

I'll appreciate any kind of feedback!

### CODE REVIEW

File: docs/CommunicationIOS.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    return (
+      <Image source={imgURI} />
+    );
+  },
+  render() {
+    return (
+      <View>
+        {this.props.imageList.map(this.renderImage)}
+      </View>
+    )
```
Comment (ide) [inline docs/CommunicationIOS.md]: semicolon

**Comment by ide** [comment]
Thanks, this looks really great and the tests all passed including website generation. I haven't read over it thoroughly yet but much appreciated. cc @nicklockwood @javache

File: docs/CommunicationIOS.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+### Limits of properties
+
+The main drawback of cross-language properties is that they do not support callbacks, which would allow us to handle bottom-up data bindings. Imagine you have a small RN view that you want to be removed from the native parent view as a result of a JS action. There is no way to do that with props, as the information would need to go bottom-up.
+
+Although we have a flavor of cross-language callbacks ([described here](https://facebook.github.io/react-native/docs/native-modules-ios.html#callbacks)), these callbacks are not always the thing we need. The main problem is that they are not intended to be passed as properties. Rather, this mechanism allows us to trigger a native action from JS, and handle the result of that action in JS.
+
+Ideally, some day we'll fully support converting JS functions to native (and vice versa), so we can pass functions transparently between native and React Native.
```
Comment (javache) [inline docs/CommunicationIOS.md]: I wouldn't put this in the documentation like this, as we're not sure we'll ever want to support this. I think most of what you describe here can be done with native modules (+ callbacks), even though it's not super flexible.

File: docs/CommunicationIOS.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+## Other ways of cross-language interaction (events and native modules)
+
+As stated in the previous chapter, using properties comes with some limitations. Sometimes properties are not enough to drive the logic of our app and we need a solution that gives more flexibility. This chapter covers other communication techniques available in React Native. They can be used for internal communication (between JS and native layers in RN) as well as for external communication (between RN and the 'pure native' part of your app).
+
+React Native supports a generic way for cross-language communication. It enables you to trigger custom functions on the other side. So you can call native functions from JS and vice versa. Unfortunately, depending on the side we are working on, we achieve the same goal in different ways. For native - we use events mechanism to schedule an execution of a handler function in JS, while for React Native we directly call methods exported by native modules.
```
Comment (javache) [inline docs/CommunicationIOS.md]: This is not very clear to me. If it's generic, why are there limitations? Could you rephrase this?

File: docs/CommunicationIOS.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+### Calling React Native functions from native (events)
+
+Events are described in details in [this article](http://facebook.github.io/react-native/docs/native-components-ios.html#events). Note that using events gives us no guarantees about execution time, as the event is handled on a separate thread.
```
Comment (javache) [inline docs/CommunicationIOS.md]: nit: in detail

File: docs/CommunicationIOS.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+Events are powerful, because they allow us to change React Native components without needing a reference to them. However, there are some pitfalls that you can fall into while using them:
+
+* As events can be send from anywhere, they can introduce spaghetti-style dependencies into your project.
```
Comment (javache) [inline docs/CommunicationIOS.md]: nit: can be sent

File: docs/CommunicationIOS.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+### Calling React Native functions from native (events)
+
+Events are described in details in [this article](http://facebook.github.io/react-native/docs/native-components-ios.html#events). Note that using events gives us no guarantees about execution time, as the event is handled on a separate thread.
+
+Events are powerful, because they allow us to change React Native components without needing a reference to them. However, there are some pitfalls that you can fall into while using them:
+
+* As events can be send from anywhere, they can introduce spaghetti-style dependencies into your project. 
+* Events share namespace, which means that you amy encounter some name collisions. Collisions will not be detected statically, what makes them hard to debug.
```
Comment (javache) [inline docs/CommunicationIOS.md]: nit: may encounter name colleisions

File: docs/CommunicationIOS.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+### Calling React Native functions from native (events)
+
+Events are described in details in [this article](http://facebook.github.io/react-native/docs/native-components-ios.html#events). Note that using events gives us no guarantees about execution time, as the event is handled on a separate thread.
+
+Events are powerful, because they allow us to change React Native components without needing a reference to them. However, there are some pitfalls that you can fall into while using them:
+
+* As events can be send from anywhere, they can introduce spaghetti-style dependencies into your project. 
+* Events share namespace, which means that you amy encounter some name collisions. Collisions will not be detected statically, what makes them hard to debug. 
+* If you use several instances of the same React Native component and you want to distinguish them from the perspective of your event, you'll likely need to introduce some kind of identifiers and pass them along with events.
```
Comment (javache) [inline docs/CommunicationIOS.md]: Perhaps suggest using the native view's reactTag as the identifier?

File: docs/CommunicationIOS.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+The common pattern we use when embedding native in React Native is to make the native component's RCTViewManager a delegate for the views, sending events back to JavaScript via the bridge. This keeps related event calls in one place.
+
+### Calling native functions from React Native (native modules)
+
+Native modules are singleton obj-c classes that are available in JS. They can export arbitrary functions and constants to React Native. They have been covered in detail in [this article](https://facebook.github.io/react-native/docs/native-modules-ios.html#content).
```
Comment (javache) [inline docs/CommunicationIOS.md]: They're not strictly singleton: typically one instance is created per JS bridge.

Comment (javache) [inline docs/CommunicationIOS.md]: nit: write out obj-c fully: Objective-C

File: docs/CommunicationIOS.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+> All native modules share the same namespace. Watch out for name collisions when creating new ones.
+
+## Layout computation flow
+
+When integrating native and React Native, we also need a way to consolidate two different layout systems. This section covers common layouting problems and provides a brief description of mechanisms that are intended to address them.
+
+### Layout of a native component embedded in React Native
+
+This case is covered in [this article](https://facebook.github.io/react-native/docs/native-components-ios.html#styles). Basically, as all our native react views are subclasses of `UIView`, most style and size attributes will work like you would expect out of the box.
+
```
Comment (javache) [inline docs/CommunicationIOS.md]: Not necessarily on you to fix, but this (or the linked documentation) needs a section on how to handle components with dynamic height using custom measure functions. Maybe @majak can help with that?

File: docs/CommunicationIOS.md [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+var React = require('react-native');
+  var {
+  View,
+  Image
+} = React;
+
+class ImageBrowserApp extends React.Component {
+  renderImage: function(imgURI) {
+    return (
+      <Image source={imgURI} />
```
Comment (nicklockwood) [inline docs/CommunicationIOS.md]: I think this needs to be:

    <Image source={{uri: imgURI}} />
